Resiliency suite for PDS and first Test Case

dbhatnagar-px opened the pull request:
Checking in Resiliency Suite along with a test case of rebooting a node during deployment. 
1. This Test case is basic framework for Resiliency like how we want to achieve resiliency execution in context of PDS
2. It validates if this is a resiliency test case
3. Based out of "n" different type of supported failures, this test case picks one failure (as of now it's just one)
4. It checks if that particular failure condition is met
5. Once failure condition is met, this test case induces that particular failure
6. It checks for any errors and fails the TC if there is any failure that happens

Have run this again the test scenario where we want to reboot a node onto which a deployment pod is coming up. Following are the exact steps this test case has: 
1. Create a Data Service
2. Check if data service is coming up 
3. Select a node onto which a pod is yet to come up
4. Reboot that node
5. After rebooting, validate the data service still comes up fine 
6. Validate configuration of data service

Jenkins Run: https://jenkins.pwx.dev.purestorage.com/job/PDS/job/pds-branch-build-test/147/

Madanagopal19 commented on tests/pds/dataservice_test.go:
```diff
@@ -312,6 +353,7 @@ var _ = Describe("{ScaleUPDataServices}", func() {
 				}
 				Step("Deploy and validate data service", func() {
 					isDeploymentsDeleted = false
+					pdslib.MarkResiliencyTC(true)
```
Is this change Intentional ?

dbhatnagar-px replied:
Removed.

Madanagopal19 commented on tests/pds/dataservice_test.go:
```diff
@@ -298,6 +298,47 @@ var _ = Describe("{EnableandDisableNamespace}", func() {
+					failure := pdslib.ResiliencyFailure{
+						Type: active_node_reboot_during_deployment,
+						Method: func() error {
+							return pdslib.RebootActiveNodeDuringDeployment(params.InfraToTest.Namespace)
```
Incase of error in this method, are we not failing it immediately?

dbhatnagar-px replied:
As discussed, it is failing immediately. Call to function and error catching is happening later than this line

Madanagopal19 commented on tests/pds/dataservice_test.go:
```diff
@@ -298,6 +298,47 @@ var _ = Describe("{EnableandDisableNamespace}", func() {
+							return pdslib.RebootActiveNodeDuringDeployment(params.InfraToTest.Namespace)
+						},
+					}
+					pdslib.DefineFailureType(failure)
```
I can see `DefineFailureType ` is updating the var `FailureType ` and `FailureType ` is not consumed by any other places. Is this expected ?

dbhatnagar-px replied:
Failuretype was the struct that was used to call it's failure method. I have updated the names now as we discussed

Madanagopal19 commented on drivers/pds/lib/utils.go:
```diff
@@ -742,6 +761,28 @@ func ValidateDataServiceDeployment(deployment *pds.ModelsDeployment, namespace s
+		}
+		if ss.Status.Replicas >= check_till_replica {
+			hasResiliencyConditionMet = true
+			log.Infof("Resiliency Condition Met. Will go ahead and try to induce failure now")
```
lets have log.InfoD

Madanagopal19 commented on drivers/pds/lib/resiliency_utils.go:
```diff
@@ -0,0 +1,125 @@
+}
+
+// Wrapper to Define failure type from Test Case
+func DefineFailureType(failure ResiliencyFailure) {
```
Lets also have a new name instead of `DefineFailureType`

dbhatnagar-px commented on drivers/pds/lib/utils.go:
```diff
@@ -742,6 +761,28 @@ func ValidateDataServiceDeployment(deployment *pds.ModelsDeployment, namespace s
+// Function to check for set amount of Replica Pods
+func GetPdsSs(depName string, ns string, check_till_replica int32) {
+	var ss *v1.StatefulSet
+	err = wait.Poll(resiliencyInterval, timeOut, func() (bool, error) {
```
Not Reqd now with channel changes

dbhatnagar-px commented on drivers/pds/lib/utils.go:
```diff
@@ -713,6 +714,24 @@ func ValidateDataServiceDeployment(deployment *pds.ModelsDeployment, namespace s
 		log.Errorf("An Error Occured while getting statefulsets %v", err)
 		return err
 	}
+	if ResiliencyFlag {
```
This comment is not reqd now

dbhatnagar-px commented:
Have addressed review comments except one - As of now it's only running for postgres for minimal testing. Once PR is ready to merge, I will simply remove that check and merge it in. 
Latest Jenkins run after addressing all review comments and also design change to incorporate channels instead of flags : 
https://jenkins.pwx.dev.purestorage.com/job/PDS/job/pds-branch-build-test/154/

dbhatnagar-px commented on tests/pds/dataservice_test.go:
```diff
@@ -298,6 +298,47 @@ var _ = Describe("{EnableandDisableNamespace}", func() {
+	It("deploy Dataservices", func() {
+		Step("Deploy Data Services", func() {
+			for _, ds := range params.DataServiceToTest {
+				if ds.Name != postgresql {
```
Keeping this comment as unresolved for now - this is for minimal testing. Will merge the PR after removing this check only.

dbhatnagar-px replied:
Resolving this. Ready for final merge.

Madanagopal19 reviewed (approved):
LGTM

Madanagopal19 commented on tests/pds/resiliency_tests.go:
```diff
@@ -0,0 +1,56 @@
+					// Deploy and Validate this Data service after injecting the type of failure we want to catch
+					deployment, _, _, err = TriggerDeployDataService(ds, params.InfraToTest.Namespace, tenantID, projectID)
+					log.FailOnError(err, "Error while deploying data services")
+					err = pdslib.InduceFailureAfterWaitingForCondition(deployment, namespace)
```
as discussed pass the replicas as params

dbhatnagar-px replied:
Done

Madanagopal19 reviewed (changes requested):
Minor changes

gauraa2 reviewed (approved):
LGTM

gauraa2 reviewed (approved):
LGTM, please fix the build issue.

lsrinivas-pure commented on drivers/pds/lib/resiliency_utils.go:
```diff
@@ -0,0 +1,165 @@
+)
+
+const (
+	active_node_reboot_during_deployment = "active-node-reboot-during-deployment"
```
please use camelcase format

dbhatnagar-px replied:
Done

lsrinivas-pure commented on drivers/pds/lib/resiliency_utils.go:
```diff
@@ -0,0 +1,165 @@
+	hasResiliencyConditionMet = false
+	FailureType               TypeOfFailure
+	testError                 error
+	check_till_replica        int32
```
please use camcelcase format

dbhatnagar-px replied:
Done

lsrinivas-pure commented on drivers/pds/lib/utils.go:
```diff
@@ -164,6 +169,7 @@ const (
 	timeOut               = 30 * time.Minute
 	timeInterval          = 10 * time.Second
 	maxtimeInterval       = 30 * time.Second
+	resiliencyInterval    = 5 * time.Second
 	envDsVersion          = "DS_VERSION"
```
I think 5 seconds is long time for replicas to be up, Is 1 second ok?

lsrinivas-pure commented on tests/pds/resiliency_tests.go:
```diff
@@ -0,0 +1,56 @@
+	It("deploy Dataservices", func() {
+		Step("Deploy Data Services", func() {
+			for _, ds := range params.DataServiceToTest {
+				if ds.Name != postgresql {
```
any reason why this test is limited to postgres?

dbhatnagar-px replied:
This is intentional for testing purposes. I will remove this check while doing the final merge. It will run for all data services.

lsrinivas-pure commented on drivers/pds/lib/resiliency_utils.go:
```diff
@@ -0,0 +1,165 @@
+
+// Close all open Resiliency channels here
+func CloseResiliencyChannel() {
+	close(ResiliencyCondition)
```
can you add some checks here to close the channel gracefully without causing panic

dbhatnagar-px replied:
Done. Although it's not required as of now as we are calling close during Post Test Execution, but it's better to have checks. Thanks for pointing it out - we will keep this in mind during further test case addition

lsrinivas-pure commented on drivers/pds/lib/resiliency_utils.go:
```diff
@@ -0,0 +1,165 @@
+}
+
+// Reboot the Active Node onto which the application pod is coming up
+func RebootActiveNodeDuringDeployment(ns string) error {
```
are you validating this error  any where?

dbhatnagar-px replied:
Yes, this is getting validated when all functions execute in parallel and proper error is being returned to the test case

lsrinivas-pure commented on drivers/pds/lib/utils.go:
```diff
@@ -786,6 +793,25 @@ func ValidateDataServiceDeployment(deployment *pds.ModelsDeployment, namespace s
+// Function to check for set amount of Replica Pods
+func GetPdsSs(depName string, ns string, check_till_replica int32) {
+	var ss *v1.StatefulSet
+	err = wait.Poll(resiliencyInterval, timeOut, func() (bool, error) {
```
return this err and handle it

dbhatnagar-px replied:
Done

Madanagopal19 reviewed (approved):
lgtm

lsrinivas-pure commented on drivers/pds/lib/resiliency_utils.go:
```diff
@@ -0,0 +1,174 @@
+func InduceFailure(failure string, ns string) {
+	isResiliencyConditionset := <-ResiliencyCondition
+	if isResiliencyConditionset {
+		FailureType.Method()
```
will this not return error?

dbhatnagar-px replied:
It will be a channel based return from next test (PR) onwards

lsrinivas-pure commented on drivers/pds/lib/resiliency_utils.go:
```diff
@@ -0,0 +1,174 @@
+		checkTillReplica = CheckTillReplica
+		log.InfoD("Entering to check if Data service has %v active pods. Once it does, we will reboot the node it is hosted upon.", checkTillReplica)
+		func1 := func() {
+			GetPdsSs(deployment.GetClusterResourceName(), namespace, checkTillReplica)
```
how are you handling the error returned here?

dbhatnagar-px replied:
It's getting handled as a separate error check condition once the go routine exits.

lsrinivas-pure commented on drivers/pds/lib/resiliency_utils.go:
```diff
@@ -0,0 +1,174 @@
+	for _, fn := range functions {
+		go func(FuncToRun func()) {
+			defer wg.Done()
+			FuncToRun()
```
if any function returns error, how are you handling it?

lsrinivas-pure reviewed (approved):
lgtm
